### PR TITLE
dma support 16550 uart

### PR DESCRIPTION
**Pull request by zhang-2019**

### CODE REVIEW

**Review by jerpelea — APPROVED** [review]

File: drivers/serial/uart_16550.c [hunk 1]
```diff
@@ -56,9 +62,24 @@
...
+  int32_t                dmarx;
+  FAR struct dma_chan_s *chanrx;
+  FAR void              *dmarxbuf;
+  size_t                 dmarxsize;
+  volatile size_t        dmarxhead;
+  volatile size_t        dmarxtail;
+  int32_t                dmarxtimeout;
+#endif
+#if !defined(CONFIG_16550_SUPRESS_CONFIG) || defined(HAVE_16550_UART_DMA)
   uint32_t         baud;      /* Configured baud */
```
Comment (pkarashchenko) [inline drivers/serial/uart_16550.c:77]: I think we need to add some spaces to align with above.

File: drivers/serial/uart_16550.c [hunk 2]
```diff
@@ -1157,25 +1317,198 @@ static bool u16550_rxflowcontrol(struct uart_dev_s *dev,
...
+      if (priv->chanrx == NULL)
+        {
+          return; /* Fail to get DMA channel */
+        }
+      u16550_dmarxconfig(dev);
+
+      /* Start a never stop DMA cyclic transfer in the background */
+      DMA_START_CYCLIC(priv->chanrx, u16550_dmareceive_done, dev,
+                       up_addrenv_va_to_pa(priv->dmarxbuf),
+                       up_addrenv_va_to_pa((void *)priv->uartbase),
```
Comment (pkarashchenko) [inline drivers/serial/uart_16550.c:1461]:
```suggestion
                       up_addrenv_va_to_pa((FAR void *)priv->uartbase),
```

File: include/nuttx/dma/dma.h [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+#define DMA_MEM_TO_MEM          1
+#define DMA_MEM_TO_DEV          2
+#define DMA_DEV_TO_MEM          3
+#define DMA_DEV_TO_DEV          4
+
+#ifdef CONFIG_DMA_LINK
+# define DMA_BLOCK_MODE         0
+# define DMA_SRC_LINK_MODE      1
+# define DMA_DST_LINK_MODE      2
+# define DMA_DUAL_LINK_MODE     3
```
Comment (pkarashchenko) [inline include/nuttx/dma/dma.h:62]:
```suggestion
#  define DMA_BLOCK_MODE         0
#  define DMA_SRC_LINK_MODE      1
#  define DMA_DST_LINK_MODE      2
#  define DMA_DUAL_LINK_MODE     3
```

**Review by pkarashchenko — COMMENTED** [review]
Let's merge https://github.com/apache/nuttx/pull/8773 first

**Comment by xiaoxiang781216** [comment]
@zhang-2019 please remove dma.h related change from the patch, which was merged here: https://github.com/apache/nuttx/pull/8773.

**Comment by acassis** [comment]
@zhang-2019 please fix this conflict

**Comment by zhang-2019** [comment]
I have resolved all the confilcts, please review

File: drivers/serial/uart_16550.c [hunk 4]
```diff
@@ -1157,25 +1317,198 @@ static bool u16550_rxflowcontrol(struct uart_dev_s *dev,
...
-#ifdef CONFIG_SERIAL_RXDMA
 static void u16550_dmareceive(FAR struct uart_dev_s *dev)
 {
+  FAR struct u16550_s *priv = dev->priv;
+
+  if (priv->dmarxhead != priv->dmarxtail)
+    {
+      size_t length = priv->dmarxhead - priv->dmarxtail;
+      size_t offset = priv->dmarxtail % priv->dmarxsize;
+      FAR void *buffer = priv->dmarxbuf + offset;
```
[the remark this answers is not in the document]
Comment (zhang-2019) [reply drivers/serial/uart_16550.c]: change the type of dmarxbuf from void * to char * would be ok

**Comment by xiaoxiang781216** [comment]
> I have resolved all the confilcts, please review

please fetch the last main and rebase your change to remove the merge commit from PR.

**Comment by zhang-2019** [comment]
I have rebased to the last commit, please review

**Review by pkarashchenko — APPROVED** [review]

File: drivers/serial/uart_16550.c [hunk 5]
```diff
@@ -160,9 +177,24 @@ static char g_uart3txbuffer[CONFIG_16550_UART3_TXBUFSIZE];
 static struct u16550_s g_uart0priv =
 {
   .uartbase       = CONFIG_16550_UART0_BASE,
-#ifndef CONFIG_16550_SUPRESS_CONFIG
+#ifdef CONFIG_16550_UART0_DMA
+  .dmatx          = CONFIG_16550_UART0_DMA_TX,
+  .dmarx          = CONFIG_16550_UART0_DMA_RX,
+#  if CONFIG_16550_UART0_DMA_RX != -1
+  .dmarxbuf       = g_uart0dmarxbuf,
```
Comment (xiaoxiang781216) [inline drivers/serial/uart_16550.c:184]: where we defefine g_uart0dmarxbuf?

**Pull request closed without merging**